### PR TITLE
Avoid non-blocking GPU->CPU copies.

### LINKED ISSUE #11287: GPU prediction race condition results in BasePredictionWriter observing incorrect (zero) values
## 🐛 Bug

When writing prediction results using a trivial BasePredictionWriter subclass with write_interval "epoch", the final batch is written as zeros when the following conditions are met:

* GPU accelerator (haven't tested with other non-CPU accelerators)
* Sufficiently expensive model that the CUDA stream has not completed by the time the prediction writer is called
  * (This is why BoringModel is not appropriate to reproduce here.)

I believe that this is a race condition resulting from non-blocking copies from GPU to CPU without explicit CUDA synchronization. This results in CPU tensors being incorrectly observed as zeros before the GPU computation completes. Details below.

### To Reproduce

```
import argparse
import logging
import os

from torchvision.datasets import FakeData
from torchvision.models import resnet50
from torchvision.transforms import ToTensor

import pytorch_lightning as pl
from pytorch_lightning.plugins import DDPSpawnPlugin
from pytorch_lightning.callbacks import BasePredictionWriter
from torch.utils.data import DataLoader
import torch


logging.basicConfig(
    format='%(asctime)s %(levelname)-8s %(message)s',
    level=logging.INFO,
    datefmt='%Y-%m-%d %H:%M:%S')


class Model(pl.LightningModule):

    def __init__(self):
        super().__init__()
        # Arbitrary model, but this is a race, and the race needs the model to
        # be expensive enough to cause the CUDA stream to queue.
        self.model = resnet50(pretrained=False)

    def forward(self, x):
        x = self.model(x)
        # Return scalar values (easier to print), and ensure that 0s are not possible.
        # It's important that the result depends on the slow self.model(x) step.
        return (x.abs() + 0.1).sum(dim=1)

    def predict_step(self, batch, batch_idx, dataloader_idx=None):
        x, _ = batch
        y = self(x)
        return y


class InferenceWriter(BasePredictionWriter):
    """Writes predictions on epoch end."""

    def __init__(self, output_path: str):
        super().__init__("epoch")
        self.output_file = os.path.join(output_path, "predictions.pt")

    def write_on_epoch_end(
        self, trainer, module, predictions, batch_indices
    ):
        assert len(predictions) == 1
        predictions = predictions[0]
        outputs = torch.cat(predictions)
        logging.info("Saving output to %s.", self.output_file)
        torch.save(outputs, self.output_file)

    def read(self):
        return torch.load(self.output_file)


parser = argparse.ArgumentParser()
parser.add_argument("--output_path", required=True)
parser.add_argument("--batch_size", default=256, type=int)
parser.add_argument("--image_size", default=288, type=int, help="Image size for inference")


def main(args):
    dataset = FakeData(
        args.batch_size * 2,
        image_size=(3, args.image_size, args.image_size),
        transform=ToTensor(),
    )
    model = Model()
    dataloader = DataLoader(
        dataset,
        batch_size=args.batch_size,
        num_workers=2,
        persistent_workers=True,  # unnecessary here, but silences warning
    )
    writer = InferenceWriter(args.output_path)
    trainer = pl.Trainer(
        devices=1,
        num_nodes=1,
        accelerator="gpu",
        default_root_dir=args.output_path,
        strategy=DDPSpawnPlugin(find_unused_parameters=False),
        callbacks=[writer],
    )
    logging.info("Starting inference")
    trainer.predict(model, dataloaders=dataloader)
    logging.info("Loading features")
    outputs = writer.read()

    logging.info("First batch values: %s", outputs[:args.batch_size])
    logging.info("Second batch values: %s", outputs[args.batch_size:])


if __name__ == '__main__':
    args = parser.parse_args()
    main(args)
```

Output:
```
2021-12-31 05:27:38 INFO     First batch values: tensor([6145.3164, 6103.9824, 6139.4873, 6153.1626, 6133.7827, 6134.3999, ...
2021-12-31 05:27:39 INFO     Second batch values: tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., ...
```

### Expected behavior

Reproduce script: Both the first and the second batch (logged in the script above) should contain nonzero values, from the same distribution. The computation is such that zeroes are not possible outputs.

In general: Tensors returned to the user on API surfaces should contain valid values when their contents are observed.

### Cause (I think)

If I understand the semantics of tensors copied from GPU to CPU with non_blocking=True, these tensors are unsafe without explicit CUDA synchronization to ensure that the CPU tensors have valid values (assuming the copy to CPU is covered by something like torch.cuda.synchronize()).

A non_blocking GPU -> CPU transfer creates a CPU tensor that is effectively a future, similar to CUDA tensors whose values have not resolved. But unlike CUDA tensors, the CPU tensor does not act like a future, and does not know that the value has not been written. As a result, operations that observe the tensor values do not wait for the tensor to be valid, and incorrect (zero) values are observed.

Since these futures are sharp edges, I suggest that we don't expose them to users unless we synchronize appropriately.

Similar bugs resulting from non-blocking GPU -> CPU copies have been reported elsewhere, for instance there are two similar reports on [this thread](https://discuss.pytorch.org/t/should-we-set-non-blocking-to-true/38234/17).

### Possible fixes

I think this could be done using cuda synchronization before callbacks that might see outputs that have been non-blocking moved to CPU. Calling `torch.cuda.synchronize()` before exposing affected tensors (CPU tensors from non-blocking copies from GPU operations) to the user might fix this.

However I think it would be better to not set non_blocking=True when copying tensors to CPU to avoid this case. It will be hard to find all the possible cases above. I don't see a strong argument to using non-blocking copies from GPU, and avoiding incorrect results due to the surprising behavior of these tensors seems like a compelling argument against. I expect that changing `move_data_to_device` to not set non_blocking when device is CPU should work.

### Fixes I've tested

Copying tensors to CPU in `predict_step`: `y = y.cpu()`. This uses a blocking transfer, making Lightning's non-blocking `.to()` call a no-op.

Inspecting the result of GPU operations also fixes this, by forcing us to wait for the GPU. This has to be done on the GPU tensors, before the CPU copy. (eg. `y.mean().item()` in `predict_step` in the reproduce example).

Using CPU compute also fixes this, but is slow, since the examples need enough computation to expose the race with GPUs.

Additionally, various things to "win" the race condition here also work:
* A simple sleep in write_on_epoch_end "fixes" this
* Using a smaller inference size (reducing GPU compute) "fixes" this

### Environment

* CUDA:
        - GPU:
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
                - A100-SXM4-40GB
        - available:         True
        - version:           11.2
* Packages:
        - numpy:             1.21.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0
        - pytorch-lightning: 1.5.7
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.9

cc @tchaton @rohitgr7

### PR DESCRIPTION
**Pull request by edpizzi**
Non-blocking GPU->CPU transfers can create race windows where tensor contents are observed to have incorrect values.

Fixes #11287. See #11287 for detailed mechanics of the issue this resolves.

Verified that this change fixes the behavior in the reproduce script from #11287.

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
   - see discussion on the issue thread for challenges with contbuild tests for this. I did test that this resolves the reproduce script in #11287
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
  - can't seem to do this, and the review guidelines wiki page no longer loads for me

## Did you have fun?

Fixing: yes. Root causing the bug: no. It's subtle, and many log statements you add to debug the issue fix the issue (by observing the value of GPU tensors, forcing a CUDA sync), which is frustrating if you don't know that this is the nature of the issue.

I also had some trouble getting tests to run in my environment. It looks to me like some tests fail in GPU environments but pass in non-GPU environments. Others fail when run on slurm environments, which is the only way I can easily access GPUs.

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -35,6 +35,9 @@
     Batch = type(None)
 
 
+CPU_DEVICES = [None, "cpu", torch.device("cpu")]
```
Comment (justusschock) [inline pytorch_lightning/utilities/apply_func.py:38]:
```suggestion
CPU_DEVICES = (None, "cpu", torch.device("cpu"))
```

Constants should not be mutable :)

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -371,6 +371,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed dataloaders not getting reloaded the correct amount of times when setting `reload_dataloaders_every_n_epochs` and `check_val_every_n_epoch` ([#10948](https://github.com/PyTorchLightning/pytorch-lightning/pull/10948))
 
+- Fixed a race that could result in incorrect (zero) values being observed in prediction writer callbacks. ([#11287](https://github.com/PyTorchLightning/pytorch-lightning/pull/11287))
```
Comment (rohitgr7) [inline CHANGELOG.md:374]:
```suggestion
- Fixed a race that could result in incorrect (zero) values being observed in prediction writer callbacks. ([#11288](https://github.com/PyTorchLightning/pytorch-lightning/pull/11288))
```

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/apply_func.py [hunk 3]
```diff
@@ -35,6 +35,9 @@
     Batch = type(None)
 
 
+CPU_DEVICES = (None, "cpu", torch.device("cpu"))
```
Comment (awaelchli) [inline pytorch_lightning/utilities/apply_func.py:38]: the device can't be None, see the type annotation below.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/apply_func.py (hunk 3 above)
Comment (edpizzi) [reply pytorch_lightning/utilities/apply_func.py:38]: My first attempt at fixing this was basically `torch.device(device) == torch.device("cpu")`, and it failed unit tests because we get None here, contrary to the type annotation.

That said, we can remove `None` here, since that case is a no-op (`torch.to(None, non_blocking=True)` is safe, and does nothing).

Comment (edpizzi) [reply pytorch_lightning/utilities/apply_func.py]: I've removed `None`, so I think this can be resolved.

File: pytorch_lightning/utilities/apply_func.py [hunk 4]
```diff
@@ -35,6 +35,9 @@
     Batch = type(None)
 
 
+CPU_DEVICES = ("cpu", torch.device("cpu"))
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:38]: This can be protected

```suggestion
_CPU_DEVICES = ("cpu", torch.device("cpu"))
```

File: pytorch_lightning/utilities/apply_func.py [hunk 5]
```diff
@@ -274,7 +277,10 @@ def batch_to(data: Any) -> Any:
                 setattr(device_data, field, device_field)
             return device_data
 
-        kwargs = dict(non_blocking=True) if isinstance(data, torch.Tensor) else {}
+        kwargs = {}
+        # Don't issue non-blocking transfers to CPU. #11287
+        if isinstance(data, torch.Tensor) and device not in CPU_DEVICES:
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:282]:
```suggestion
        if isinstance(data, torch.Tensor) and device not in _CPU_DEVICES:
```

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -274,7 +277,10 @@ def batch_to(data: Any) -> Any:
                 setattr(device_data, field, device_field)
             return device_data
 
-        kwargs = dict(non_blocking=True) if isinstance(data, torch.Tensor) else {}
+        kwargs = {}
+        # Don't issue non-blocking transfers to CPU. #11287
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:281]: There's git blame to find the associated PR number

```suggestion
        # Don't issue non-blocking transfers to CPU
```

**Review by carmocca — APPROVED** [review]
